### ISSUE TITLE
False positive on "Cannot move" inspection on a type with derived Copy.

### ISSUE THREAD

**Issue by alexander-irbis** [intellij-rust/intellij-rust#3101]
### 1

* **Intellij-Rust plugin version:** 0.2.0.2110-182
* **Rust toolchain version:** stable-x86_64-unknown-linux-gnu (default)
rustc 1.30.1 (1433507eb 2018-11-07)
* **IDE name and version:** IntelliJ IDEA 2018.2.6 (Community Edition)
Build #IC-182.5107.16, built on November 6, 2018
JRE: 1.8.0_152-release-1248-b22 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
* **Operating system:** Linux 4.19.1-gentoo

### 2

* **Intellij-Rust plugin version:** v0.2.0.2111-183
* **Rust toolchain version:** stable-x86_64-unknown-linux-gnu (default)
rustc 1.30.1 (1433507eb 2018-11-07)
* **IDE name and version:** IntelliJ IDEA 2018.3 (Community Edition)
Build #IC-183.4284.148, built on November 21, 2018
JRE: 1.8.0_152-release-1343-b15 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
* **Operating system:** Linux 4.19.1-gentoo

## Problem description

![screenshot_20181201_184701](https://user-images.githubusercontent.com/1448323/49331066-30607380-f5a0-11e8-977f-3ed555c920b5.png)

## Steps to reproduce

```rust
#[derive(Copy, Clone)]
struct X {}

fn copy(x: X) -> (X, X) {
    (x, x)
}
```

```rust
#[derive(Copy, Clone)]
enum X {}

fn copy(x: X) -> (X, X) {
    (x, x)
}
```

**Comment by alexander-irbis** [comment]
## Environment

### 3

* **Intellij-Rust plugin version:** v0.2.0.2111-183
* **Rust toolchain version:** stable-x86_64-apple-darwin (default)
rustc 1.30.1 (1433507eb 2018-11-07)
* **IDE name and version:** IntelliJ IDEA 2018.3 (Ultimate Edition)
Build #IU-183.4284.148, built on November 21, 2018
JRE: 1.8.0_152-release-1343-b15 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
* **Operating system:** macOS 10.14.1


On this environment I can't reproduce the issue.

**Comment by ortem** [comment]
Can't reproduce this on macOS 10.14.1 with `v0.2.0.2111-183` and on Ubuntu 4.15.0-39-generic with `v0.2.0.2111-183`.
Have you tried to delete `#[derive(Copy, Clone)]` and write once again?

**Comment by aldanor** [comment]
Here's an example that produces wrong "Cannot move" inspection on macOS (1.31 stable; latest intellij rust):

```rust
#[derive(Clone, Copy)]
struct Foo<T: Copy> {
    x: T,
}

impl<T: Copy> Foo<T> {
    fn foo(&self) -> Self {
        Self { x: self.x }  // <-- "Cannot move"
    }
}
```

**Comment by aldanor** [comment]
Here's another one:

```rust
#[derive(Clone, Copy)]
struct Foo<T: Copy> {
    x: T,
}

fn foo<T: Copy>(x: &Foo<T>) -> T {
    x.x // <-- "Cannot move"
}
```

**Comment by ortem** [comment]
@aldanor This problem has been already fixed [here](https://github.com/intellij-rust/intellij-rust/pull/3094). Just wait for the new release a little bit 😄 

Also, the original problem is not related to the move analysis and doesn't reproduce, so I close it.

**Comment by mrobakowski** [comment]
I just noticed this still happens, when you refer to derive-trait by a fully specified name. It's enough that one of the types is in its full form and then none work.

Minimal example (which, to be honest, is a little ridiculous):
```rust
#[derive(Copy, std::marker::Clone)]
struct X;
```
and a slightly more real-world example:
```rust
#[derive(prost::Message, Copy, Clone)]
struct Message {
  payload: u64,
}
```

**Comment by Undin** [comment]
@mrobakowski Thanks for the feedback!
I've created separate issue for your case - #4850
